### PR TITLE
New Post: An Intro to WebSockets

### PR DESCRIPTION
**Pull request by aksh555**
This article gives a brief history of the evolution of WebSockets and the problems which they solve.
Screenshot from 2019-10-04 17-10-19
Screenshot from 2019-10-04 17-10-24

.

### CODE REVIEW

**Comment by aksh555** [comment]
> The contents of the blog really looks good to me. Great work @aksh555

Thanks @SaurabhAgarwala.

File: blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+**AJAX should still be used for making short-lived web service calls**
+
+## Further Reading
+
+WebSockets can be implemented practically using Node.js(using Socket.IO), Rails(using Action Cable) and in other frameworks too.
+
+## References
+
+- https://medium.com/@yassimortensen/an-introduction-to-websockets-10b131182559 
+- https://blog.logrocket.com/websockets-tutorial-how-to-go-real-time-with-node-and-react-8e4693fbf843/
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md]: Add these two as links

File: blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+date: 2019-10-03 23:32:44
+image: '/assets/img/'
+description: 'Evolution of WebSockets'
+tags:
+- IEEE NITK
+- CompSoc
+- Web
+- Networks
+categories:
+- Compsoc
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md]: CompSoc

File: scripts/_site/image_optimise.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	if height > max_height:
+		new_width = ((int)((width*1.0/height)*max_height))
+		print(width/height)
+		img = img.resize((new_width,max_height), Image.ANTIALIAS)
+
+	width,height = img.size
+	print(str(width) + " " + str(height))
+
+	img.save(file_name, optimize=True, quality=75)
```
Comment (salman-bhai) [inline scripts/_site/image_optimise.py]: Umm why is this file required here?

File: scripts/_site/compress.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import os
+import PIL
+from PIL import Image
+
+for root, dirs, files in os.walk("."):  
+    for filename in files:
+    	if ".png" in filename:
+        	foo = Image.open(filename)
+        	foo.save(filename,quality=95,optimize = True)
```
Comment (salman-bhai) [inline scripts/_site/compress.py]: Umm why is this file required here?

File: scripts/_site/cibuild [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+set -e # halt script on error
+
+jekyll build
+htmlproofer ./_site --disable-external
+
+rm -r ./_site
```
Comment (salman-bhai) [inline scripts/_site/cibuild]: Same comments for this file as the others

File: blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md]: Remove extra lines please

File: blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+This went on just fine until.....
+Clients started sending more frequent requests to the server requesting for data.
+A good example of this is web apps that need real-time data (like gaming or chat apps). These require an abuse of the HTTP protocol to establish bidirectional data transfer.
+
+# Problem
+The main problem in the traditional HTTP model was that it was **always the client that initiated transactions, ie. it was unidirectional**. So a number of strategies were proposed to enable the server to 'push' data to the client.
+An other reason was increasing overhead due to HTTP. Every time a HTTP request is made, a bunch of headers and cookie data are transferred to the server. This can add up to a reasonably large amount of data that needs to be transferred, which in turn increases latency.
+
+*Basically imagine that each time to get an updated stock's price/a new message from your friend you had to refresh the page, and wait until the new price/message* *loaded. Firstly how would you even know that there has been a change in price/a new message has been sent to to you? And even if you did, it would take*
+*quite a while to load.*
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md]: Is this intended to be in different lines?

Additionally you could use `>` symbol to add it as a blockquote. Google it up once

File: blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+*quite a while to load.* 
+
+Some of the **startegies** devised to combat the issue were:
+### 1. Optimising HTTP Polling
+- **Periodic Polling:**
+The very first attempt to solve the problem was by polling the server at regular intervals. It could reduce server consumption by increasing time gap between two requests. But ofcourse notifications of server events would be delayed :/
+- **Long Polling:**
+After the server receives a request, it holds the request open until new data(if any) is available. There were no notification delays.But more server resources were used.
+- **Streaming:**
+Basically infinitely polling...But the connection could get interrupted!
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md]: Could you maybe describe how the connection could get interrupted?

**Comment by salman-bhai** [comment]
@Hrily interesting read on Websockets 😛

**Comment by salman-bhai** [comment]
Resolve Merge Conflicts @aksh555

**Comment by aksh555** [comment]
> Resolve Merge Conflicts @aksh555

It's because of .gitignore. You had told to add .jekyll-cache and .idea to it.

**Comment by salman-bhai** [comment]
> > Resolve Merge Conflicts @aksh555
> 
> @salman-bhai It's because of .gitignore. You had told to add .jekyll-cache and .idea to it.

@aksh555 yes you're right. But you need to pull from master, resolve locally else won't be able to merge your MR

File: .gitignore [hunk 9]
```diff
@@ -13,4 +13,4 @@ gyan_src/_drafts/*
 .vagrant/*
 .jekyll-cache/
 .jekyll-metadata
-.idea
+
```
Comment (salman-bhai) [inline .gitignore]: Umm don't remove this

File: blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md [hunk 10]
```diff
@@ -32,20 +32,21 @@ Clients started sending more frequent requests to the server requesting for data
...
+The main problem in the traditional HTTP model was that it was **always the client that initiated transactions, ie. it was unidirectional**. 
+An other reason was increasing overheads due to HTTP. Every time a HTTP request is made, a bunch of headers and cookie data are transferred to the server. This can add up to a reasonably large amount of data that needs to be transferred, which in turn increases latency.
 
-*Basically imagine that each time to get an updated stock's price/a new message from your friend you had to refresh the page, and wait until the new price/message* *loaded. Firstly how would you even know that there has been a change in price/a new message has been sent to to you? And even if you did, it would take*
-*quite a while to load.* 
+>Basically imagine that each time to get an updated stock's price/a new message from your friend you had to refresh the page, and wait until the new price/message loaded. Firstly how would you even know that there has been a change in price/a new message has been sent to to you? And even if you did, it would take*
```
Comment (salman-bhai) [inline blog_src/_posts/2019-10-03-An-Intro-to-WebSockets.md]: Remove the `*` at the end of the line

**Comment by MJ10** [comment]
@aksh555 can you please make the requested changes and fix the merge conflicts?

**Comment by aksh555** [comment]
> @aksh555 can you please make the requested changes and resolve the merge conflicts?

Yeah, sure

**Comment by MJ10** [comment]
@aksh555 any updates? It's fine if you make the changes in this PR itself, a new PR isn't necessary
